Static provisioning requires nonsensical managed resources

<!--
Thank you for helping to improve Crossplane!

Please be sure to search for open issues before raising a new one. We use issues
for bug reports and feature requests. Please find us at https://slack.crossplane.io
for questions, support, and discussion.
-->

### What happened?
<!--
Please let us know what behaviour you expected and how Crossplane diverged from
that behaviour.
-->
Changes were made to our resource claim controller watch predicates recently in https://github.com/crossplaneio/crossplane-runtime/pull/18 and https://github.com/crossplaneio/crossplane-runtime/pull/24. The latter refactored the watch predicates used by resource claim controllers with the intention of supporting both statically and dynamically provisioned managed resources. It does this by accepting resources that satisfy any of three watch predicates, applied in the following order:

1. `resource.HasManagedResourceReferenceKind` accepts resource claims with a `.spec.resourceRef` to the kind of managed resource the reconciler is concerned with.
1. `resource.HasDirectClassReferenceKind` accepts managed resources with a `.spec.classRef` to the kind of non-portable class the reconciler is concerned with.
1. `resource.HasIndirectClassReferenceKind` accepts resource claims with a `.spec.classRef` to a portable class with a `.classRef` to the kind of non-portable class the reconciler is concerned with.

In retrospect `resource.HasDirectClassReferenceKind` only makes sense in a dynamic provisioning context, because statically provisioned managed resources don't use a resource class, and thus typically don't set their `.spec.classRef`. Based on my read of the current predicate logic (I have not confirmed this) I expect our resource claim reconcilers would never notice changes to statically provisioned managed resources.

This is not too big an issue given the resource claim reconciler only watches managed resources to wait for them to become bindable, which they typically do before the resource claim is created in a static provisioning scenario. I'm guessing that's why I didn't encounter this case when I tested static provisioning per https://github.com/crossplaneio/crossplane-runtime/pull/24#issuecomment-531460570.

### How can we reproduce it?
<!--
Help us to reproduce your bug as succinctly and precisely as possible. Artifacts
such as example manifests or a script that triggers the issue are highly
appreciated!
-->
1. Create a resource claim that references a managed resource that does not yet exist.
1. Create the referenced managed resource.

My guess is that the claim will never bind.

There are two workarounds for this:
1. Create the managed resource and wait for it its `.status.bindingPhase` to become `Unbound` before creating a resource claim that references it.
1. Set the managed resource's `.spec.classRef.kind` and `.spec.classRef.apiVersion` to its corresponding non-portable resource class kind.

### What environment did it happen in?
crossplane-runtime version: 26a458d 

<!--
Include at least the version or commit of Crossplane you were running. Consider
also including your:

* Cloud provider or hardware configuration
* Kubernetes version (use `kubectl version`)
* Kubernetes distribution (e.g. Tectonic, GKE, OpenShift)
* OS (e.g. from /etc/os-release)
* Kernel (e.g. `uname -a`)
-->

> I'd expect scenario three to result in a nil pointer exception in the claim reconciler

Yuuup. If I submit the same resource claim before the managed resource exists I see:

```
l/wait/wait.go:88"}
E1024 01:30:33.462044   16337 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereferenc
e)
/Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:76
/Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:65
/Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:51
/usr/local/Cellar/go@1.12/1.12.10/libexec/src/runtime/panic.go:522
/usr/local/Cellar/go@1.12/1.12.10/libexec/src/runtime/panic.go:82
/usr/local/Cellar/go@1.12/1.12.10/libexec/src/runtime/signal_unix.go:390
/Users/negz/control/go/pkg/mod/github.com/crossplaneio/crossplane-runtime@v0.0.0-20191023215652-0f37bea5496e/pkg/resource/claim_binding_reconciler.go:348
/Users/negz/control/go/pkg/mod/github.com/crossplaneio/crossplane-runtime@v0.0.0-20191023215652-0f37bea5496e/pkg/resource/claim_binding_reconciler.go:348
/Users/negz/control/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0/pkg/internal/controller/controller.go:216
/Users/negz/control/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0/pkg/internal/controller/controller.go:192
/Users/negz/control/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0/pkg/internal/controller/controller.go:171
/Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152
/Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153
/Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88
/usr/local/Cellar/go@1.12/1.12.10/libexec/src/runtime/asm_amd64.s:1337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1f1b66e]

goroutine 712 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:58 +0x105
panic(0x248a6a0, 0x38f45d0)
        /usr/local/Cellar/go@1.12/1.12.10/libexec/src/runtime/panic.go:522 +0x1b5
github.com/crossplaneio/crossplane-runtime/pkg/meta.NamespacedNameOf(...)
        /Users/negz/control/go/pkg/mod/github.com/crossplaneio/crossplane-runtime@v0.0.0-20191023215652-0f37bea5496e/pkg/meta/meta.go:90
github.com/crossplaneio/crossplane-runtime/pkg/resource.(*ClaimReconciler).Reconcile(0xc000286000, 0xc000040be9, 0x7, 0xc000040bb0, 0xe, 0x390ee00, 0x0, 0x0, 0x0)
        /Users/negz/control/go/pkg/mod/github.com/crossplaneio/crossplane-runtime@v0.0.0-20191023215652-0f37bea5496e/pkg/resource/claim_binding_reconciler.go:348 +0x242e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0002b4000, 0x24f2620, 0xc000c10420, 0x24f2600)
        /Users/negz/control/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0/pkg/internal/controller/controller.go:216 +0x146
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0002b4000, 0xc000116e00)
        /Users/negz/control/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0/pkg/internal/controller/controller.go:192 +0xb5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0002b4000)
        /Users/negz/control/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0/pkg/internal/controller/controller.go:171 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00085c060)
        /Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00085c060, 0x3b9aca00, 0x0, 0x1, 0xc0001381e0)
        /Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00085c060, 0x3b9aca00, 0xc0001381e0)
        /Users/negz/control/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
        /Users/negz/control/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0/pkg/internal/controller/controller.go:157 +0x311
```

--- a/pkg/resource/api.go
+++ b/pkg/resource/api.go
@@ -147,11 +147,12 @@ func (a *APIManagedConnectionPropagator) PropagateConnection(ctx context.Context
 // using the status subresource; such objects should use APIManagedStatusBinder.
 type APIManagedBinder struct {
 	client client.Client
+	typer  runtime.ObjectTyper
 }
 
 // NewAPIManagedBinder returns a new APIManagedBinder.
-func NewAPIManagedBinder(c client.Client) *APIManagedBinder {
-	return &APIManagedBinder{client: c}
+func NewAPIManagedBinder(c client.Client, t runtime.ObjectTyper) *APIManagedBinder {
+	return &APIManagedBinder{client: c, typer: t}
 }
 
 // Bind the supplied resource to the supplied claim.
@@ -164,6 +165,12 @@ func (a *APIManagedBinder) Bind(ctx context.Context, cm Claim, mg Managed) error
 			return errors.Wrap(err, errUpdateClaim)
 		}
 	}
+
+	// This claim reference will already be set for dynamically provisioned
+	// managed resources, but we need to make sure it's set for statically
+	// provisioned resources too.
+	cmr := meta.ReferenceTo(cm, MustGetKind(cm, a.typer))
+	mg.SetClaimReference(cmr)
 	mg.SetBindingPhase(v1alpha1.BindingPhaseBound)
 	if err := a.client.Update(ctx, mg); err != nil {
 		return errors.Wrap(err, errUpdateManaged)
@@ -177,11 +184,12 @@ func (a *APIManagedBinder) Bind(ctx context.Context, cm Claim, mg Managed) error
 // APIManagedBinder.
 type APIManagedStatusBinder struct {
 	client client.Client
+	typer  runtime.ObjectTyper
 }
 
 // NewAPIManagedStatusBinder returns a new APIManagedStatusBinder.
-func NewAPIManagedStatusBinder(c client.Client) *APIManagedStatusBinder {
-	return &APIManagedStatusBinder{client: c}
+func NewAPIManagedStatusBinder(c client.Client, t runtime.ObjectTyper) *APIManagedStatusBinder {
+	return &APIManagedStatusBinder{client: c, typer: t}
 }
 
 // Bind the supplied resource to the supplied claim.
@@ -194,6 +202,16 @@ func (a *APIManagedStatusBinder) Bind(ctx context.Context, cm Claim, mg Managed)
 			return errors.Wrap(err, errUpdateClaim)
 		}
 	}
+
+	// This claim reference will already be set for dynamically provisioned
+	// managed resources, but we need to make sure it's set for statically
+	// provisioned resources too.
+	cmr := meta.ReferenceTo(cm, MustGetKind(cm, a.typer))
+	mg.SetClaimReference(cmr)
+	if err := a.client.Update(ctx, mg); err != nil {
+		return errors.Wrap(err, errUpdateManaged)
+	}
+
 	mg.SetBindingPhase(v1alpha1.BindingPhaseBound)
 	if err := a.client.Status().Update(ctx, mg); err != nil {
 		return errors.Wrap(err, errUpdateManagedStatus)

--- a/pkg/resource/api.go
+++ b/pkg/resource/api.go
@@ -158,13 +158,6 @@ func NewAPIManagedBinder(c client.Client, t runtime.ObjectTyper) *APIManagedBind
 // Bind the supplied resource to the supplied claim.
 func (a *APIManagedBinder) Bind(ctx context.Context, cm Claim, mg Managed) error {
 	cm.SetBindingPhase(v1alpha1.BindingPhaseBound)
-	// Propagate back the final name of the external resource to the claim.
-	if meta.GetExternalName(mg) != "" {
-		meta.SetExternalName(cm, meta.GetExternalName(mg))
-		if err := a.client.Update(ctx, cm); err != nil {
-			return errors.Wrap(err, errUpdateClaim)
-		}
-	}
 
 	// This claim reference will already be set for dynamically provisioned
 	// managed resources, but we need to make sure it's set for statically
@@ -175,7 +168,14 @@ func (a *APIManagedBinder) Bind(ctx context.Context, cm Claim, mg Managed) error
 	if err := a.client.Update(ctx, mg); err != nil {
 		return errors.Wrap(err, errUpdateManaged)
 	}
-	return nil
+
+	if meta.GetExternalName(mg) == "" {
+		return nil
+	}
+
+	// Propagate back the final name of the external resource to the claim.
+	meta.SetExternalName(cm, meta.GetExternalName(mg))
+	return errors.Wrap(a.client.Update(ctx, cm), errUpdateClaim)
 }
 
 // An APIManagedStatusBinder binds resources to claims by updating them in a
@@ -195,13 +195,6 @@ func NewAPIManagedStatusBinder(c client.Client, t runtime.ObjectTyper) *APIManag
 // Bind the supplied resource to the supplied claim.
 func (a *APIManagedStatusBinder) Bind(ctx context.Context, cm Claim, mg Managed) error {
 	cm.SetBindingPhase(v1alpha1.BindingPhaseBound)
-	// Propagate back the final name of the external resource to the claim.
-	if meta.GetExternalName(mg) != "" {
-		meta.SetExternalName(cm, meta.GetExternalName(mg))
-		if err := a.client.Update(ctx, cm); err != nil {
-			return errors.Wrap(err, errUpdateClaim)
-		}
-	}
 
 	// This claim reference will already be set for dynamically provisioned
 	// managed resources, but we need to make sure it's set for statically
@@ -216,7 +209,14 @@ func (a *APIManagedStatusBinder) Bind(ctx context.Context, cm Claim, mg Managed)
 	if err := a.client.Status().Update(ctx, mg); err != nil {
 		return errors.Wrap(err, errUpdateManagedStatus)
 	}
-	return nil
+
+	if meta.GetExternalName(mg) == "" {
+		return nil
+	}
+
+	// Propagate back the final name of the external resource to the claim.
+	meta.SetExternalName(cm, meta.GetExternalName(mg))
+	return errors.Wrap(a.client.Update(ctx, cm), errUpdateClaim)
 }
 
 // An APIManagedUnbinder finalizes the deletion of a managed resource by

--- a/pkg/resource/api_test.go
+++ b/pkg/resource/api_test.go
@@ -464,6 +464,7 @@ func TestBind(t *testing.T) {
 	}
 
 	errBoom := errors.New("boom")
+	externalName := "very-cool-external-resource"
 
 	cases := map[string]struct {
 		client client.Client
@@ -472,8 +473,15 @@ func TestBind(t *testing.T) {
 		want   want
 	}{
 		"UpdateManagedError": {
-			client: &test.MockClient{MockUpdate: test.NewMockUpdateFn(errBoom)},
-			typer:  MockSchemeWith(&MockClaim{}),
+			client: &test.MockClient{MockUpdate: test.NewMockUpdateFn(nil, func(obj runtime.Object) error {
+				switch obj.(type) {
+				case *MockManaged:
+					return errBoom
+				default:
+					return errors.New("unexpected object kind")
+				}
+			})},
+			typer: MockSchemeWith(&MockClaim{}),
 			args: args{
 				ctx: context.Background(),
 				cm:  &MockClaim{},
@@ -488,7 +496,39 @@ func TestBind(t *testing.T) {
 				},
 			},
 		},
-		"Successful": {
+		"UpdateClaimError": {
+			client: &test.MockClient{MockUpdate: test.NewMockUpdateFn(nil, func(obj runtime.Object) error {
+				switch obj.(type) {
+				case *MockManaged:
+					return nil
+				case *MockClaim:
+					return errBoom
+				default:
+					return errors.New("unexpected object kind")
+				}
+			})},
+			typer: MockSchemeWith(&MockClaim{}),
+			args: args{
+				ctx: context.Background(),
+				cm:  &MockClaim{},
+				mg: &MockManaged{
+					ObjectMeta: metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+				},
+			},
+			want: want{
+				err: errors.Wrap(errBoom, errUpdateClaim),
+				cm: &MockClaim{
+					ObjectMeta:   metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+					MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
+				mg: &MockManaged{
+					ObjectMeta:          metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
+					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
+			},
+		},
+		"SuccessfulWithoutExternalName": {
 			client: &test.MockClient{MockUpdate: test.NewMockUpdateFn(nil)},
 			typer:  MockSchemeWith(&MockClaim{}),
 			args: args{
@@ -500,6 +540,28 @@ func TestBind(t *testing.T) {
 				err: nil,
 				cm:  &MockClaim{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
 				mg: &MockManaged{
+					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
+					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
+			},
+		},
+		"SuccessfulWithExternalName": {
+			client: &test.MockClient{MockUpdate: test.NewMockUpdateFn(nil)},
+			typer:  MockSchemeWith(&MockClaim{}),
+			args: args{
+				ctx: context.Background(),
+				cm:  &MockClaim{},
+				mg: &MockManaged{
+					ObjectMeta: metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+				},
+			},
+			want: want{
+				err: nil,
+				cm: &MockClaim{
+					ObjectMeta:   metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+					MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
+				mg: &MockManaged{
+					ObjectMeta:          metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
 					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
 					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
 				},
@@ -538,6 +600,7 @@ func TestStatusBind(t *testing.T) {
 	}
 
 	errBoom := errors.New("boom")
+	externalName := "very-cool-external-resource"
 
 	cases := map[string]struct {
 		client client.Client
@@ -546,9 +609,14 @@ func TestStatusBind(t *testing.T) {
 		want   want
 	}{
 		"UpdateManagedError": {
-			client: &test.MockClient{
-				MockUpdate: test.NewMockUpdateFn(errBoom),
-			},
+			client: &test.MockClient{MockUpdate: test.NewMockUpdateFn(nil, func(obj runtime.Object) error {
+				switch obj.(type) {
+				case *MockManaged:
+					return errBoom
+				default:
+					return errors.New("unexpected object kind")
+				}
+			})},
 			typer: MockSchemeWith(&MockClaim{}),
 			args: args{
 				ctx: context.Background(),
@@ -583,7 +651,42 @@ func TestStatusBind(t *testing.T) {
 				},
 			},
 		},
-		"Successful": {
+		"UpdateClaimError": {
+			client: &test.MockClient{
+				MockUpdate: test.NewMockUpdateFn(nil, func(obj runtime.Object) error {
+					switch obj.(type) {
+					case *MockManaged:
+						return nil
+					case *MockClaim:
+						return errBoom
+					default:
+						return errors.New("unexpected object kind")
+					}
+				}),
+				MockStatusUpdate: test.NewMockStatusUpdateFn(nil),
+			},
+			typer: MockSchemeWith(&MockClaim{}),
+			args: args{
+				ctx: context.Background(),
+				cm:  &MockClaim{},
+				mg: &MockManaged{
+					ObjectMeta: metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+				},
+			},
+			want: want{
+				err: errors.Wrap(errBoom, errUpdateClaim),
+				cm: &MockClaim{
+					ObjectMeta:   metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+					MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
+				mg: &MockManaged{
+					ObjectMeta:          metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
+					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
+			},
+		},
+		"SuccessfulWithoutExternalName": {
 			client: &test.MockClient{
 				MockUpdate:       test.NewMockUpdateFn(nil),
 				MockStatusUpdate: test.NewMockStatusUpdateFn(nil),
@@ -598,6 +701,32 @@ func TestStatusBind(t *testing.T) {
 				err: nil,
 				cm:  &MockClaim{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
 				mg: &MockManaged{
+					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
+					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
+			},
+		},
+		"SuccessfulWithExternalName": {
+			client: &test.MockClient{
+				MockUpdate:       test.NewMockUpdateFn(nil),
+				MockStatusUpdate: test.NewMockStatusUpdateFn(nil),
+			},
+			typer: MockSchemeWith(&MockClaim{}),
+			args: args{
+				ctx: context.Background(),
+				cm:  &MockClaim{},
+				mg: &MockManaged{
+					ObjectMeta: metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+				},
+			},
+			want: want{
+				err: nil,
+				cm: &MockClaim{
+					ObjectMeta:   metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
+					MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
+				mg: &MockManaged{
+					ObjectMeta:          metav1.ObjectMeta{Annotations: map[string]string{meta.ExternalNameAnnotationKey: externalName}},
 					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
 					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
 				},

--- a/pkg/resource/api_test.go
+++ b/pkg/resource/api_test.go
@@ -467,11 +467,13 @@ func TestBind(t *testing.T) {
 
 	cases := map[string]struct {
 		client client.Client
+		typer  runtime.ObjectTyper
 		args   args
 		want   want
 	}{
 		"UpdateManagedError": {
 			client: &test.MockClient{MockUpdate: test.NewMockUpdateFn(errBoom)},
+			typer:  MockSchemeWith(&MockClaim{}),
 			args: args{
 				ctx: context.Background(),
 				cm:  &MockClaim{},
@@ -480,11 +482,15 @@ func TestBind(t *testing.T) {
 			want: want{
 				err: errors.Wrap(errBoom, errUpdateManaged),
 				cm:  &MockClaim{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
-				mg:  &MockManaged{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
+				mg: &MockManaged{
+					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
+					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
 			},
 		},
 		"Successful": {
 			client: &test.MockClient{MockUpdate: test.NewMockUpdateFn(nil)},
+			typer:  MockSchemeWith(&MockClaim{}),
 			args: args{
 				ctx: context.Background(),
 				cm:  &MockClaim{},
@@ -493,14 +499,17 @@ func TestBind(t *testing.T) {
 			want: want{
 				err: nil,
 				cm:  &MockClaim{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
-				mg:  &MockManaged{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
+				mg: &MockManaged{
+					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
+					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
 			},
 		},
 	}
 
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
-			api := NewAPIManagedBinder(tc.client)
+			api := NewAPIManagedBinder(tc.client, tc.typer)
 			err := api.Bind(tc.args.ctx, tc.args.cm, tc.args.mg)
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
 				t.Errorf("api.Bind(...): -want error, +got error:\n%s", diff)
@@ -532,11 +541,34 @@ func TestStatusBind(t *testing.T) {
 
 	cases := map[string]struct {
 		client client.Client
+		typer  runtime.ObjectTyper
 		args   args
 		want   want
 	}{
+		"UpdateManagedError": {
+			client: &test.MockClient{
+				MockUpdate: test.NewMockUpdateFn(errBoom),
+			},
+			typer: MockSchemeWith(&MockClaim{}),
+			args: args{
+				ctx: context.Background(),
+				cm:  &MockClaim{},
+				mg:  &MockManaged{},
+			},
+			want: want{
+				err: errors.Wrap(errBoom, errUpdateManaged),
+				cm:  &MockClaim{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
+				mg: &MockManaged{
+					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
+				},
+			},
+		},
 		"UpdateManagedStatusError": {
-			client: &test.MockClient{MockStatusUpdate: test.NewMockStatusUpdateFn(errBoom)},
+			client: &test.MockClient{
+				MockUpdate:       test.NewMockUpdateFn(nil),
+				MockStatusUpdate: test.NewMockStatusUpdateFn(errBoom),
+			},
+			typer: MockSchemeWith(&MockClaim{}),
 			args: args{
 				ctx: context.Background(),
 				cm:  &MockClaim{},
@@ -545,11 +577,18 @@ func TestStatusBind(t *testing.T) {
 			want: want{
 				err: errors.Wrap(errBoom, errUpdateManagedStatus),
 				cm:  &MockClaim{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
-				mg:  &MockManaged{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
+				mg: &MockManaged{
+					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
+					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
 			},
 		},
 		"Successful": {
-			client: &test.MockClient{MockStatusUpdate: test.NewMockStatusUpdateFn(nil)},
+			client: &test.MockClient{
+				MockUpdate:       test.NewMockUpdateFn(nil),
+				MockStatusUpdate: test.NewMockStatusUpdateFn(nil),
+			},
+			typer: MockSchemeWith(&MockClaim{}),
 			args: args{
 				ctx: context.Background(),
 				cm:  &MockClaim{},
@@ -558,14 +597,17 @@ func TestStatusBind(t *testing.T) {
 			want: want{
 				err: nil,
 				cm:  &MockClaim{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
-				mg:  &MockManaged{MockBindable: MockBindable{Phase: v1alpha1.BindingPhaseBound}},
+				mg: &MockManaged{
+					MockClaimReferencer: MockClaimReferencer{meta.ReferenceTo(&MockClaim{}, MockGVK(&MockClaim{}))},
+					MockBindable:        MockBindable{Phase: v1alpha1.BindingPhaseBound},
+				},
 			},
 		},
 	}
 
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
-			api := NewAPIManagedStatusBinder(tc.client)
+			api := NewAPIManagedStatusBinder(tc.client, tc.typer)
 			err := api.Bind(tc.args.ctx, tc.args.cm, tc.args.mg)
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
 				t.Errorf("api.Bind(...): -want error, +got error:\n%s", diff)

--- a/pkg/resource/claim_binding_reconciler.go
+++ b/pkg/resource/claim_binding_reconciler.go
@@ -351,9 +351,14 @@ func (r *ClaimReconciler) Reconcile(req reconcile.Request) (reconcile.Result, er
 		return reconcile.Result{Requeue: false}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errUpdateClaimStatus)
 	}
 
-	if !meta.WasCreated(managed) {
-		// Class reference should always be set by the time we get this far;
-		// Our watch predicates require it.
+	// Claim reconcilers (should) watch for either claims with a resource ref,
+	// claims with a class ref, or managed resources with a claim ref. In the
+	// first case the managed resource always exists by the time we get here. In
+	// the second case the class reference is set. The third case exposes us to
+	// a pathological scenario in which a managed resource references a claim
+	// that has no resource ref or class ref, so we can't assume the class ref
+	// is always set at this point.
+	if !meta.WasCreated(managed) && claim.GetClassReference() != nil {
 		class := r.newClass()
 		// Class reference should always be set by the time we get this far; we
 		// set it on last reconciliation.

--- a/pkg/resource/claim_binding_reconciler.go
+++ b/pkg/resource/claim_binding_reconciler.go
@@ -191,7 +191,7 @@ func defaultCRManaged(m manager.Manager) crManaged {
 		ManagedConfigurator:         NewObjectMetaConfigurator(m.GetScheme()),
 		ManagedCreator:              NewAPIManagedCreator(m.GetClient(), m.GetScheme()),
 		ManagedConnectionPropagator: NewAPIManagedConnectionPropagator(m.GetClient(), m.GetScheme()),
-		ManagedBinder:               NewAPIManagedBinder(m.GetClient()),
+		ManagedBinder:               NewAPIManagedBinder(m.GetClient(), m.GetScheme()),
 		ManagedFinalizer:            NewAPIManagedUnbinder(m.GetClient()),
 	}
 }
@@ -289,7 +289,7 @@ func NewClaimReconciler(m manager.Manager, of ClaimKind, using ClassKind, with M
 
 // Reconcile a resource claim with a concrete managed resource.
 func (r *ClaimReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
-	// NOTE(negz): This method is a little over our cyclomatic complexity goal.
+	// NOTE(negz): This method is well over our cyclomatic complexity goal.
 	// Be wary of adding additional complexity.
 
 	log.V(logging.Debug).Info("Reconciling", "controller", claimControllerName, "request", req)
@@ -390,6 +390,16 @@ func (r *ClaimReconciler) Reconcile(req reconcile.Request) (reconcile.Result, er
 	}
 
 	if !IsBindable(managed) && !IsBound(managed) {
+		if managed.GetClaimReference() == nil {
+			// We're waiting to bind to a statically provisioned managed
+			// resource. We must requeue because our EnqueueRequestForClaim
+			// handler can only enqueue reconciles for managed resource updates
+			// when they have their claim reference set, and that doesn't happen
+			// until we bind to the managed resource we're waiting for.
+			claim.SetConditions(Binding(), v1alpha1.ReconcileSuccess())
+			return reconcile.Result{RequeueAfter: aShortWait}, errors.Wrap(r.client.Status().Update(ctx, claim), errUpdateClaimStatus)
+		}
+
 		// If this claim was not already binding we'll be requeued due to the
 		// status update. Otherwise there's no need to requeue. We should be
 		// watching both the resource claims and the resources we own, so we'll

--- a/pkg/resource/claim_binding_reconciler.go
+++ b/pkg/resource/claim_binding_reconciler.go
@@ -22,6 +22,7 @@ import (
 
 	"github.com/pkg/errors"
 	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"sigs.k8s.io/controller-runtime/pkg/client"
@@ -305,7 +306,18 @@ func (r *ClaimReconciler) Reconcile(req reconcile.Request) (reconcile.Result, er
 
 	managed := r.newManaged()
 	if ref := claim.GetResourceReference(); ref != nil {
-		if err := IgnoreNotFound(r.client.Get(ctx, meta.NamespacedNameOf(ref), managed)); err != nil {
+		err := r.client.Get(ctx, meta.NamespacedNameOf(ref), managed)
+		if kerrors.IsNotFound(err) {
+			// If the managed resource we explicitly reference doesn't exist yet
+			// we want to retry after a brief wait, in case it is created. We
+			// must explicitly requeue because our EnqueueRequestForClaim
+			// handler can only enqueue reconciles for managed resources that
+			// have their claim reference set, so we can't expect to be queued
+			// implicitly when the managed resource we want to bind to appears.
+			claim.SetConditions(Binding(), v1alpha1.ReconcileSuccess())
+			return reconcile.Result{RequeueAfter: aShortWait}, errors.Wrap(r.client.Status().Update(ctx, claim), errUpdateClaimStatus)
+		}
+		if err != nil {
 			// If we didn't hit this error last time we'll be requeued
 			// implicitly due to the status update. Otherwise we want to retry
 			// after a brief wait, in case this was a transient error.

--- a/pkg/resource/claim_binding_reconciler_test.go
+++ b/pkg/resource/claim_binding_reconciler_test.go
@@ -24,6 +24,7 @@ import (
 	"github.com/google/go-cmp/cmp"
 	"github.com/pkg/errors"
 	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
@@ -114,6 +115,41 @@ func TestClaimReconciler(t *testing.T) {
 							want := &MockClaim{}
 							want.SetResourceReference(&corev1.ObjectReference{})
 							want.SetConditions(v1alpha1.ReconcileError(errBoom))
+							if diff := cmp.Diff(want, got, test.EquateConditions()); diff != "" {
+								t.Errorf("-want, +got:\n%s", diff)
+							}
+							return nil
+						}),
+					},
+					s: MockSchemeWith(&MockClaim{}, &MockClass{}, &MockManaged{}),
+				},
+				of:   ClaimKind(MockGVK(&MockClaim{})),
+				use:  ClassKind(MockGVK(&MockClass{})),
+				with: ManagedKind(MockGVK(&MockManaged{})),
+			},
+			want: want{result: reconcile.Result{RequeueAfter: aShortWait}},
+		},
+		"ManagedNotFound": {
+			args: args{
+				m: &MockManager{
+					c: &test.MockClient{
+						MockGet: test.NewMockGetFn(nil, func(o runtime.Object) error {
+							switch o := o.(type) {
+							case *MockClaim:
+								cm := &MockClaim{}
+								cm.SetResourceReference(&corev1.ObjectReference{})
+								*o = *cm
+								return nil
+							case *MockManaged:
+								return kerrors.NewNotFound(schema.GroupResource{}, "")
+							default:
+								return errUnexpected
+							}
+						}),
+						MockStatusUpdate: test.NewMockStatusUpdateFn(nil, func(got runtime.Object) error {
+							want := &MockClaim{}
+							want.SetResourceReference(&corev1.ObjectReference{})
+							want.SetConditions(Binding(), v1alpha1.ReconcileSuccess())
 							if diff := cmp.Diff(want, got, test.EquateConditions()); diff != "" {
 								t.Errorf("-want, +got:\n%s", diff)
 							}

--- a/pkg/resource/claim_binding_reconciler_test.go
+++ b/pkg/resource/claim_binding_reconciler_test.go
@@ -311,6 +311,38 @@ func TestClaimReconciler(t *testing.T) {
 			},
 			want: want{result: reconcile.Result{Requeue: false}},
 		},
+		"ClassReferenceNotSet": {
+			args: args{
+				m: &MockManager{
+					c: &test.MockClient{
+						MockGet: test.NewMockGetFn(nil, func(o runtime.Object) error {
+							switch o := o.(type) {
+							case *MockClaim:
+								*o = MockClaim{}
+								return nil
+							case *MockManaged:
+								return nil
+							default:
+								return errUnexpected
+							}
+						}),
+						MockStatusUpdate: test.NewMockStatusUpdateFn(nil, func(got runtime.Object) error {
+							want := &MockClaim{}
+							want.SetConditions(Binding(), v1alpha1.ReconcileSuccess())
+							if diff := cmp.Diff(want, got, test.EquateConditions()); diff != "" {
+								t.Errorf("-want, +got:\n%s", diff)
+							}
+							return nil
+						}),
+					},
+					s: MockSchemeWith(&MockClaim{}, &MockClass{}, &MockManaged{}),
+				},
+				of:   ClaimKind(MockGVK(&MockClaim{})),
+				use:  ClassKind(MockGVK(&MockClass{})),
+				with: ManagedKind(MockGVK(&MockManaged{})),
+			},
+			want: want{result: reconcile.Result{Requeue: false}},
+		},
 		"GetResourceClassError": {
 			args: args{
 				m: &MockManager{

--- a/pkg/resource/claim_binding_reconciler_test.go
+++ b/pkg/resource/claim_binding_reconciler_test.go
@@ -341,7 +341,7 @@ func TestClaimReconciler(t *testing.T) {
 				use:  ClassKind(MockGVK(&MockClass{})),
 				with: ManagedKind(MockGVK(&MockManaged{})),
 			},
-			want: want{result: reconcile.Result{Requeue: false}},
+			want: want{result: reconcile.Result{RequeueAfter: aShortWait}},
 		},
 		"GetResourceClassError": {
 			args: args{
@@ -475,6 +475,7 @@ func TestClaimReconciler(t *testing.T) {
 								// because the zero value of BindingPhase is
 								// BindingPhaseUnset.
 								mg := &MockManaged{}
+								mg.SetClaimReference(&corev1.ObjectReference{})
 								mg.SetCreationTimestamp(now)
 								*o = *mg
 								return nil
@@ -514,6 +515,7 @@ func TestClaimReconciler(t *testing.T) {
 							case *MockManaged:
 								mg := &MockManaged{}
 								mg.SetCreationTimestamp(now)
+								mg.SetClaimReference(&corev1.ObjectReference{})
 								mg.SetBindingPhase(v1alpha1.BindingPhaseUnbindable)
 								*o = *mg
 								return nil

